### PR TITLE
ArC: initial documentation of CDI pitfalls with reactive programming

### PR DESCRIPTION
**Pull request by Ladicek**

### CODE REVIEW

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 1]
```diff
@@ -1069,6 +1069,46 @@ public class NoopAsyncObserverExceptionHandler implements AsyncObserverException
...
+[[reactive_pitfalls]]
+== CDI Pitfalls with Reactive Programming
+
+CDI is a purely synchronous framework.
+Its notion of asynchrony is very limited and based solely on thread pools and thread offloading.
+Therefore, there is a number of pitfalls when using CDI together with reactive programming.
+
+=== Detecting When Blocking Is Allowed
+
+The `BlockingOperationControl.isBlockingAllowed()` method allows detecting when blocking is allowed.
```
Comment (mkouba) [inline docs/src/main/asciidoc/cdi-reference.adoc:1081]:
```suggestion
The `io.quarkus.runtime.BlockingOperationControl#isBlockingAllowed()` method can be used to detect whether blocking is allowed on the current thread.
```

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 2]
```diff
@@ -1069,6 +1069,46 @@ public class NoopAsyncObserverExceptionHandler implements AsyncObserverException
...
+When it is not, and you need to perform a blocking operation, you have to offload it to another thread.
```
Comment (mkouba) [inline docs/src/main/asciidoc/cdi-reference.adoc:1082]: Maybe we should give an example of a blocking operation or add a link to Quarkus/external docs?

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 3]
```diff
@@ -1069,6 +1069,46 @@ public class NoopAsyncObserverExceptionHandler implements AsyncObserverException
 }
 ----
 
+[[reactive_pitfalls]]
+== CDI Pitfalls with Reactive Programming
```
Comment (mkouba) [inline docs/src/main/asciidoc/cdi-reference.adoc:1073]:
```suggestion
== Pitfalls with Reactive Programming
```

**Review by mkouba — COMMENTED** [review]
This is a good start!

File: docs/src/main/asciidoc/cdi-reference.adoc (hunk 2 above)
Comment (Ladicek) [reply docs/src/main/asciidoc/cdi-reference.adoc:1082]: Maybe best to do that in the source code listing? Say, just `Thread.sleep(5000);`?

Comment (mkouba) [reply docs/src/main/asciidoc/cdi-reference.adoc:1082]: Yes, maybe with a comment like:
```java
// Write data to the filesystem, use JDBC, call a REST service, etc.
Thread.sleep(5000);
```

Comment (Ladicek) [reply docs/src/main/asciidoc/cdi-reference.adoc:1082]: Done.

**Review by manovotn — APPROVED** [review]
Good write-up.
I'd merge it as-is and expand on it as we encounter more cases in the future.

**Review by mkouba — APPROVED** [review]

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 4]
```diff
@@ -1069,6 +1069,47 @@ public class NoopAsyncObserverExceptionHandler implements AsyncObserverException
 }
 ----
 
+[[reactive_pitfalls]]
+== Pitfalls with Reactive Programming
+
+CDI is a purely synchronous framework.
+Its notion of asynchrony is very limited and based solely on thread pools and thread offloading.
```
Comment (maxandersen) [inline docs/src/main/asciidoc/cdi-reference.adoc:1076]: is that the case for async event/observers ?

Comment (Ladicek) [reply docs/src/main/asciidoc/cdi-reference.adoc:1076]: Yes. That's the only place in CDI where asynchrony is mentioned, and it's based on thread offloading.

**Pull request merged**